services/peer/api.go

```diff
@@ -0,0 +1,48 @@
+package peer
+
+import (
+	"context"
+	"errors"
+)
+
+var (
+	// ErrInvalidTopic error returned when the requested topic is not valid.
+	ErrInvalidTopic = errors.New("Topic not valid")
+
+	// ErrInvalidRange error returned when max-min range is not valid.
+	ErrInvalidRange = errors.New("Invalid range, Min should be lower or equal to Max")
+
+	// ErrDiscovererNotProvided error when discoverer is not being provided.
+	ErrDiscovererNotProvided = errors.New("Discoverer not provided")
+)
+
+// PublicAPI represents a set of APIs from the `web3.peer` namespace.
+type PublicAPI struct {
+	s *Service
+}
+
+// NewAPI creates an instance of the peer API.
+func NewAPI(s *Service) *PublicAPI {
+	return &PublicAPI{s: s}
+}
+
+// DiscoverRequest json request for peer_discover.
+type DiscoverRequest struct {
+	Topic string `json:"topic"`
+	Max   int    `json:"max"`
+	Min   int    `json:"min"`
+}
+
+// Discover is an implementation of `peer_discover` or `web3.peer.discover` API.
+func (api *PublicAPI) Discover(context context.Context, req DiscoverRequest) (err error) {
+	if len(req.Topic) == 0 {
+		return ErrInvalidTopic
+	}
+	if req.Max < req.Min {
+		return ErrInvalidRange
+	}
+	if api.s.d == nil {
```

Comment:
I would check it first. If `Discoverer` is `nil`, there is no point to check request attributes.

Reply:
Also, if you do that, then it will be easier to write unit tests. There would be two test functions needed:
1. To test if it returns a proper error if `Discoverer` is not set,
2. A table-driven test to check other errors for the request attributes verification.